src/middleware/index.ts

```diff
@@ -1,16 +1,19 @@
 import { UiState, InstantSearch } from '../types';
 
-export type MiddlewareDefinition = {
+export type MiddlewareDefinition<TExtraDefinition> = {
+  $$type: string;
   onStateChange(options: { uiState: UiState }): void;
   subscribe(): void;
   unsubscribe(): void;
-};
+} & TExtraDefinition;
```

Comment:
`insights` middleware needs to have something extra. What do you think of the usage of generic here? Any better way to achieve it?

Reply:
is this still needed?

Reply:
not anymore